File: http/http-client-annotation-processor/src/main/java/ru/tinkoff/kora/http/client/annotation/processor/extension/HttpClientLinkerExtensionFactory.java
package ru.tinkoff.kora.http.client.annotation.processor.extension;

import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionFactory;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static ru.tinkoff.kora.http.client.annotation.processor.HttpClientClassNames.httpClientAnnotation;

public class HttpClientLinkerExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var httpClient = processingEnvironment.getElementUtils().getTypeElement(httpClientAnnotation.canonicalName());
        if (httpClient == null) {
            return Optional.empty();
        } else {
            return Optional.of(new HttpClientKoraExtension(processingEnvironment));
        }
    }
}

File: http/http-client-symbol-processor/src/main/kotlin/ru/tinkoff/kora/http/client/symbol/processor/extension/HttpClientLinkerExtensionFactory.kt
package ru.tinkoff.kora.http.client.symbol.processor.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.kora.http.client.symbol.processor.HttpClientClassNames.httpClientAnnotation
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionFactory
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension

class HttpClientLinkerExtensionFactory : ExtensionFactory {

    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        if (resolver.getClassDeclarationByName(httpClientAnnotation.canonicalName) != null) {
            return HttpClientKoraExtension()
        }
        return null
    }
}
